### ISSUE TITLE
Add cross-language translation data

### ISSUE THREAD

**Issue by andrewtavis** [scribe-org/Scribe-Data#23]
### Terms

- [X] I have searched [open and closed data issues](https://github.com/scribe-org/Scribe-Data/issues?q=is%3Aissue+label%3Adata+)
- [X] I agree to follow Scribe-Data's [Code of Conduct](https://github.com/scribe-org/Scribe-Data/blob/main/.github/CODE_OF_CONDUCT.md)

### Languages

All languages

### Description

This issue is the interim step to adding full translation support to Scribe apps. What it entails is finding accurate enough models on [🤗 Hugging Face](https://huggingface.co/) to translate between all currently supported languages and English. The `format_translations.py` script for each language will then need to be edited to run each model over a basic corpus to generate seven different `translations.json` files per language.

From there this will allow an option for which base language to translate from to be added to Scribe-iOS' menu, which will be developed in https://github.com/scribe-org/Scribe-iOS/issues/16. 

- Completing this issue will unblock https://github.com/scribe-org/Scribe-iOS/issues/7, as users will be able to decide which language to translate from

**Comment by andrewtavis** [comment]
Included in this issue is:
- [ ] An exploration of running the translation models on [Google Colab](https://colab.research.google.com/) or another platform that will give us GPU access in order to speed the model runs up
- [ ] Updating the readmes with translation ranges of the least and most translations available

**Comment by andrewtavis** [comment]
The following are the language models from [🤗 Hugging Face](https://huggingface.co/) that we can use to generate the translations (checked when implemented):

- [ ] English <-> French:
- [ ] English <-> German:
- [ ] English <-> Italian:
- [ ] English <-> Portuguese:
- [ ] English <-> Russian:
- [ ] English <-> Spanish:
- [ ] English <-> Swedish:
- [ ] French <-> German:
- [ ] French <-> German:
- [ ] French <-> German:
- [ ] French <-> German:
- [ ] French <-> German:
- [ ] French <-> German:
- [ ] German <-> Italian:
- [ ] German <-> Italian:
- [ ] German <-> Italian:
- [ ] German <-> Italian:
- [ ] German <-> Italian:
- [ ] Italian <-> Portuguese:
- [ ] Italian <-> Portuguese:
- [ ] Italian <-> Portuguese:
- [ ] Italian <-> Portuguese:
- [ ] Portuguese <-> Russian:
- [ ] Portuguese <-> Russian:
- [ ] Portuguese <-> Russian:

**Comment by abhijeet78880** [comment]
hey..👋🏻
@andrewtavis  I have gone through the issue and done some research regarding this . but can you please explain more about this.

**Comment by andrewtavis** [comment]
Hey @abhijeet78880 👋 Yes I definitely expected that more information would be needed. This is a tough one, but as I said just needs some persistence 😊

Will write more later today! :)

**Comment by andrewtavis** [comment]
Ok, sooooo :) There's plenty to do here, with the first part of it being some research you/we could do. [This comment](https://github.com/scribe-org/Scribe-Data/issues/23#issuecomment-1304587756) is an ongoing list of all machine translation models available from [🤗 Hugging Face](https://huggingface.co/) that we're using. As of now we only translate from English to the language that a user is typing in, but we want to expand this so that the user can translate from any of Scribe's supported languages to any other. This will be an option in the new menu we'll build in [this iOS issue](https://github.com/scribe-org/Scribe-iOS/issues/16), with the designs for that being found [here](https://www.figma.com/file/c8945w2iyoPYVhsqW7vRn6/scribe_public_designs?node-id=405%3A393&t=zvww3pdVGdjWVmwp-1).

Once we have a translation model we make a file like [this one that translates from English to German](https://github.com/scribe-org/Scribe-Data/blob/main/src/scribe_data/extract_transform/German/translations/format_translations.py). In it we load a JSON that's words that we query in the source language (for now English) from [Wikidata](https://wikidata.org/), then we make a list of words from the JSON, set the `model` variable with the model we're documenting in [the comment above that I already mentioned](https://github.com/scribe-org/Scribe-Data/issues/23#issuecomment-1304587756), and then in the for loop we go through and translate the words and add them all into a new JSON. This JSON is then used in Scribe apps to provide a translation 😊 It's not perfect, but it's what we're doing for now 🙃

What would be great is if you/we could look and find the missing models that we need for the new translations. A lot of these can just be `Helsinki-NLP/opus-mt` models, which are giant machine translation (hence `mt`) models from researchers in Helsinki. As we're using [Helsinki-NLP/opus-mt-en-fr](https://huggingface.co/Helsinki-NLP/opus-mt-en-fr) for English to French, we can use [Helsinki-NLP/opus-mt-fr-en](https://huggingface.co/Helsinki-NLP/opus-mt-fr-en) for French to English 🚀

At this point I think it's best if I check in with you and see how the above sounds. If you want to contribute in a simple way at first, going through and finding links for what models we'd need would be best. I could then do a more in depth explanation of the [translation file](https://github.com/scribe-org/Scribe-Data/blob/main/src/scribe_data/extract_transform/German/translations/format_translations.py) from before and show you how to set up new ones for each of the models we find. We'd then run them, and bam we'd have translation data that we'd then reference and thus give users the option to translate from Spanish to German and all the other options 😊

This has been a lot! Again, just let me know how it sounds and feel free to ask questions. Everything we write here will help make things easier going forward :) :)

**Comment by andrewtavis** [comment]
@abhijeet78880, FYI I also made #35 just now which might be a nice first issue for you 😊 That's checking if the word `Scribe` is in the database for a given language and adding it if not :)

**Comment by andrewtavis** [comment]
Note that I've updated the [models comment](https://github.com/scribe-org/Scribe-Data/issues/23#issuecomment-1304587756) with further models we could use from Helsinki-NLP. There does appear to be some holes in their translation model coverage, so for some pairs we'll need to look harder for other models.

At this point we'd be ready to start copying over some translation files 😊 For this the script from any Helsinki-NLP translation file can be used and the model name just needs to be changed :)

**Comment by andrewtavis** [comment]
The following two models could help us plug some of the holes in the above translation coverage:

- https://huggingface.co/alirezamsh/small100
- https://huggingface.co/docs/transformers/model_doc/marian

**Comment by andrewtavis** [comment]
Neither of the above models was what we were looking for, and after playing around with [T5](https://huggingface.co/docs/transformers/model_doc/t5) a bit more and getting some very sub par German-Portuguese translations I was able to get some strong results from a dummy JSON dataset using [m2m100_418M](https://huggingface.co/facebook/m2m100_418M). I'd say that the small model is enough for our purposes as the single word or small phrase translation that we're doing isn't going to be improved by a larger model (or only marginally, as larger models would be taking advantage of contextual information that our short input strings would lack). 

[m2m100_418M](https://huggingface.co/facebook/m2m100_418M) should be able to handle all the missing language pairs for Scribe keyboard languages. A general thought might be to create a data pipeline that would use it as the sole model and then just switch the inputs and output languages as well as the input data during the run. Another thing to factor is that for now the outputs I was getting were capitalized as I'm assuming the model is expecting a sentence. This can be remedied by the metadata that comes from Wikidata though, as we'll be querying a base translation corpus and would thus know if a word is a proper noun and would thus need to be capitalized (or all nouns in German), or just lower case it.

Will continue to fine tune the current example and then present the results at the next Scribe Weekly 😊

**Comment by nyfz18** [comment]
Hi! Sorry for the delay -- I had to figure out a bunch of stuff on my end. Where should I start?

**Comment by andrewtavis** [comment]
No stress on a delay, @nyfz18! Sorry for mine as well :) Let me organize some stuff and I'll send along some pseudocode for how this would be written as I said I would 😊 Generally the steps would be:

- I'll write some SPARQL scripts to get words from each language
  - We'll call those scripts to update the data each time before the translation's are done
- I'll write a base script so that we can pass a language to the process like `'English'`
  - This allows us to test it easier
- We then add the following steps:
  - For any language that we're passing or all languages if we haven't past something
    - Run the script to update the data for this language (source language)
    - Load in the appropriate translation model
    - Translate from the source language to all the other languages
    - As we translate we want to update a dictionary
    - Save this dictionary to a JSON
    - Continue to the next language (if necessary)

Do you have any questions on the above, @nyfz18?

**Comment by nyfz18** [comment]
Okay, sounds good. I sort of understand, but a check in call might be more helpful!

**Comment by andrewtavis** [comment]
Hey there @nyfz18! You now have [extract_transform/translate.py](https://github.com/scribe-org/Scribe-Data/blob/main/src/scribe_data/extract_transform/translate.py) at your disposal that loads in the model, checks arguments if you'd like to pass them, and prints out the ISO codes at the end. Following the working code there's also some pseudocode that outlines the steps we discussed in the call 😊 Let me know if you have any questions/comments!

**Comment by andrewtavis** [comment]
@nyfz18, we'll be doing the conversion of the JSON data that's being produced here in the new issue #46. @lillian-mo will work on that one 🚀

**Comment by andrewtavis** [comment]
Closing this issue as individual ones have been made for each language that can be worked on as a part of Google Summer of Code ☀️ Thanks all for the discussion here! Help on the individual issues would be welcome 😊
